snapcraft/internal/build_providers/_snap.py

```diff
@@ -0,0 +1,296 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2018 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import contextlib
+import datetime
+import enum
+import logging
+import os
+import tempfile
+from typing import List
+from typing import Any, Dict, Union  # noqa: F401
+
+import yaml
+
+from . import errors
+from snapcraft.internal import repo
+
+logger = logging.getLogger(__name__)
+
+
+class _SnapOp(enum.Enum):
+    NOP = 0
+    INJECT = 1
+    INSTALL = 2
+    REFRESH = 3
+
+
+class _SnapManager:
+    def __init__(self, *, snap_name: str, snap_dir: str, latest_revision: str) -> None:
+        self.snap_name = snap_name
+        self._snap_dir = snap_dir
+
+        self._latest_revision = latest_revision
+        self.__required_operation = None  # type: Union[None, _SnapOp]
+        self.__repo = None  # type: Union[None, repo.snaps.SnapPackage]
+        self.__revision = None  # type: Union[None, str]
+        self.__install_cmd = None  # type: Union[None, List[str]]
+
+    def _get_snap_repo(self):
+        if self.__repo is None:
+            self.__repo = repo.snaps.SnapPackage(self.snap_name)
+        return self.__repo
+
+    def get_op(self) -> _SnapOp:
+        if self.__required_operation is not None:
+            return self.__required_operation
+
+        snap_repo = self._get_snap_repo()
+
+        # The latest revision is used in two ways:
+        # - if there is no snapcraft on the host, used to determine if
+        #   a refresh is needed (instead of a fresh install).
+        # - if there is a snapcraft on the host, used to determine if
+        #   re-injecting is required.
+        snap_info = snap_repo.get_local_snap_info()
+
+        if not snap_repo.installed and self._latest_revision is None:
+            op = _SnapOp.INSTALL
+        elif not snap_repo.installed and self._latest_revision is not None:
+            op = _SnapOp.REFRESH
```

Comment:
I'm a little confused on this one-- how can a refresh be what's needed if it's not installed?

Reply:
If you are running from the deb, it will not be installed, if the environment has already been created it may have a revision installed.
snap_repo is for the host, the operation for the build environment.

Reply:
I will document with code comments